### PR TITLE
Default Picking Slot Set by Workstation

### PR DESCRIPTION
**Pull request by dm-ps**
refs: https://github.com/metasfresh/metasfresh/issues/18476

### CODE REVIEW

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/service/commands/PickingJobAllocatePickingSlotCommand.java [hunk 1]
```diff
@@ -6,33 +6,45 @@
...
+import javax.annotation.Nullable;
+import java.util.Optional;
+
 public class PickingJobAllocatePickingSlotCommand
 {
 	private final ITrxManager trxManager = Services.get(ITrxManager.class);
-	@NonNull final PickingJobRepository pickingJobRepository;
-	@NonNull private final PickingJobSlotService pickingSlotService;
+	@NonNull
+	final PickingJobRepository pickingJobRepository;
```
Comment (cp-ps) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/service/commands/PickingJobAllocatePickingSlotCommand.java:22]: can be private

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/workplace/interceptor/C_Workplace.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		if (pickingSlotId == null)
+		{
+			return;
+		}
+
+		final I_M_PickingSlot pickingSlot = pickingSlotBL.getById(pickingSlotId);
+
+		if (pickingSlot.getM_Warehouse_ID() != workplace.getM_Warehouse_ID())
+		{
+			throw new AdempiereException(TranslatableStrings.builder()
```
Comment (cp-ps) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/workplace/interceptor/C_Workplace.java:72]: use plain string or actual translatable ones

**Review by cp-ps — CHANGES_REQUESTED** [review]

**Review by cp-ps — APPROVED** [review]

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/workplace/Workplace.java [hunk 3]
```diff
@@ -35,5 +36,6 @@ public class Workplace
 {
 	@NonNull WorkplaceId id;
 	@NonNull String name;
-	@Nullable WarehouseId warehouseId;
+	@NonNull WarehouseId warehouseId;
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/workplace/Workplace.java:39]: why making warehouse mandatory?

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/service/commands/PickingJobCreateCommand.java [hunk 4]
```diff
@@ -136,6 +143,36 @@ private PickingJob executeInTrx()
 		}
 	}
 
+	@NonNull
+	private PickingJob allocatePickingSlotIfPossible(@NonNull final PickingJob pickingJob)
+	{
+		final PickingSlotId pickingSlotId = loadingSupportServices.getWorkplaceByUserId(request.getPickerId())
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/service/commands/PickingJobCreateCommand.java:149]: why we introduced getWorkplaceByUserId to loading support service
and why not keeping that logic here?

asking because IMHO getWorkplaceByUserId is not related to "loading services"

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/service/commands/PickingJobCreateCommand.java [hunk 5]
```diff
@@ -136,6 +143,36 @@ private PickingJob executeInTrx()
...
+			return PickingJobAllocatePickingSlotCommand.builder()
+					.pickingJobRepository(pickingJobRepository)
+					.pickingSlotService(pickingJobSlotService)
+					//
+					.pickingJob(pickingJob)
+					.pickingSlotId(pickingSlotId)
+					//
+					.build().execute();
+		}
+		catch (final Exception allocatePickingSlotException)
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/service/commands/PickingJobCreateCommand.java:169]: pls avoid flow-control via exceptions

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/repository/DefaultPickingJobLoaderSupportingServices.java [hunk 6]
```diff
@@ -129,6 +136,13 @@ public PickingSlotIdAndCaption getPickingSlotIdAndCaption(@NonNull final Picking
 		return pickingSlotIdAndCaptionsCache.computeIfAbsent(pickingSlotId, pickingSlotService::getPickingSlotIdAndCaption);
 	}
 
+	@Override
+	@NonNull
+	public Optional<Workplace> getWorkplaceByUserId(@NonNull final UserId userId)
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/job/repository/DefaultPickingJobLoaderSupportingServices.java:141]: IMHO this is not related to loading ....

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/workplace/interceptor/C_Workplace.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+	public C_Workplace()
+	{
+		Services.get(IProgramaticCalloutProvider.class).registerAnnotatedCallout(this);
+	}
+
+	@CalloutMethod(columnNames = I_C_Workplace.COLUMNNAME_M_Warehouse_ID)
+	public void resetPickingSlotId(@NonNull final I_C_Workplace workplace)
+	{
+		workplace.setM_PickingSlot_ID(-1);
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/workplace/interceptor/C_Workplace.java:55]: isn't this happening out of the box?

**Review by teosarca — COMMENTED** [review]
@dm-ps @cp-ps pls check my comments

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/workplace/Workplace.java (hunk 3 above)
Comment (dm-ps) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/workplace/Workplace.java:39]: It is mandatory in the database

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/workplace/interceptor/C_Workplace.java (hunk 7 above)
Comment (dm-ps) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/workplace/interceptor/C_Workplace.java:55]: Not sure what you mean by out of the box.
If I remove this and try to change the warehouse in UI, you'll see an error message, coming from the validator just below this method

Comment (teosarca) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/workplace/interceptor/C_Workplace.java:55]: @dm-ps  
I suppose we have an AD_Val_Rule which filters out the picking slots based on currently selected C_Workplace.M_Warehouse_ID

In that case, when you change the C_Workplace.M_Warehouse_ID,
I expect the C_Workplace.M_PickingSlot_ID field to be validated and because the current value is no longer valid, the field to be reset.

If it's not working like that, I am a bit puzzled because we relly on this behavior in many places.
So, in that case, a workaround is NOT an option.
